### ISSUE TITLE
poetry.bash-completion has multiple syntax errors

### ISSUE THREAD

**Issue by pythoninthegrass** [python-poetry/poetry#4572]
- [x] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) Poetry version.
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option).

- **OS version and name**: macOS 10.14.6 Mojave
- **Poetry version**: Poetry (version 1.2.0a2)
- **Link of a [Gist](https://gist.github.com/) with the contents of your pyproject.toml file**: N/A

## Issue
Generated a completion file for bash via `poetry completions bash > /usr/local/etc/profile.d/poetry.bash-completion` due to `poetry completions bash > /etc/bash_completion.d/poetry.bash-completion` not having the `/etc/bash_completion.d` directory on Mac. Sourced in `.bashrc` and wound up with a bunch of syntax errors that I had to comment out (i.e., not use.)

## `.bashrc`
`. /usr/local/etc/profile.d/poetry.bash-completion`

## `/usr/local/etc/profile.d/poetry.bash-completion`
```bash
-bash: /usr/local/etc/profile.d/poetry.bash-completion: line 40: syntax error near unexpected token `clear'
-bash: /usr/local/etc/profile.d/poetry.bash-completion: line 40: `            (cache clear)'
$ which poetry
/usr/local/bin/poetry
$ source ~/.bashrc
-bash: /usr/local/etc/profile.d/poetry.bash-completion: line 44: syntax error near unexpected token `list'
-bash: /usr/local/etc/profile.d/poetry.bash-completion: line 44: `            (cache list)'
$ source ~/.bashrc
-bash: /usr/local/etc/profile.d/poetry.bash-completion: line 56: syntax error near unexpected token `info'
-bash: /usr/local/etc/profile.d/poetry.bash-completion: line 56: `            (debug info)'
$ source ~/.bashrc
-bash: /usr/local/etc/profile.d/poetry.bash-completion: line 60: syntax error near unexpected token `resolve'
-bash: /usr/local/etc/profile.d/poetry.bash-completion: line 60: `            (debug resolve)'
$ source ~/.bashrc
-bash: /usr/local/etc/profile.d/poetry.bash-completion: line 64: syntax error near unexpected token `info'
-bash: /usr/local/etc/profile.d/poetry.bash-completion: line 64: `            (env info)'
$ source ~/.bashrc
-bash: /usr/local/etc/profile.d/poetry.bash-completion: line 68: syntax error near unexpected token `list'
-bash: /usr/local/etc/profile.d/poetry.bash-completion: line 68: `            (env list)'
$ source ~/.bashrc
-bash: /usr/local/etc/profile.d/poetry.bash-completion: line 72: syntax error near unexpected token `remove'
-bash: /usr/local/etc/profile.d/poetry.bash-completion: line 72: `            (env remove)'
$ source ~/.bashrc
-bash: /usr/local/etc/profile.d/poetry.bash-completion: line 76: syntax error near unexpected token `use'
-bash: /usr/local/etc/profile.d/poetry.bash-completion: line 76: `            (env use)'
$ source ~/.bashrc
-bash: /usr/local/etc/profile.d/poetry.bash-completion: line 108: syntax error near unexpected token `add'
-bash: /usr/local/etc/profile.d/poetry.bash-completion: line 108: `            (plugin add)'
$ source ~/.bashrc
-bash: /usr/local/etc/profile.d/poetry.bash-completion: line 112: syntax error near unexpected token `remove'
-bash: /usr/local/etc/profile.d/poetry.bash-completion: line 112: `            (plugin remove)'
$ !!
source ~/.bashrc
-bash: /usr/local/etc/profile.d/poetry.bash-completion: line 116: syntax error near unexpected token `show'
-bash: /usr/local/etc/profile.d/poetry.bash-completion: line 116: `            (plugin show)'
$ !!
source ~/.bashrc
-bash: /usr/local/etc/profile.d/poetry.bash-completion: line 136: syntax error near unexpected token `update'
-bash: /usr/local/etc/profile.d/poetry.bash-completion: line 136: `            (self update)'
$
$ !!
source ~/.bashrc
-bash: /usr/local/etc/profile.d/poetry.bash-completion: line 148: syntax error near unexpected token `add'
-bash: /usr/local/etc/profile.d/poetry.bash-completion: line 148: `            (source add)'
$ !!
source ~/.bashrc
-bash: /usr/local/etc/profile.d/poetry.bash-completion: line 152: syntax error near unexpected token `remove'
-bash: /usr/local/etc/profile.d/poetry.bash-completion: line 152: `            (source remove)'
$ !!
source ~/.bashrc
-bash: /usr/local/etc/profile.d/poetry.bash-completion: line 156: syntax error near unexpected token `show'
-bash: /usr/local/etc/profile.d/poetry.bash-completion: line 156: `            (source show)'
$ !!
source ~/.bashrc
$
```
Full [poetry.bash-completion](https://gist.github.com/pythoninthegrass/da79d8a21c46875e0ac085ed4f04ff9c)

**Comment by pythoninthegrass** [comment]
Attached the generated completion file with commented out lines.

**Comment by sgaist** [comment]
Hi,

Can you add the version of bash you are using and how you installed Poetry and bash ?

I have tested your file with bash 5.1.8 on High Sierra and it worked properly.

**Comment by pythoninthegrass** [comment]
`bash` was installed via `brew` and `poetry` is `pip`.
```bash
$ bash --version
GNU bash, version 5.0.11(1)-release (x86_64-apple-darwin18.6.0)
Copyright (C) 2019 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
 $ pip list
Package            Version
------------------ ---------
...
poetry             1.2.0a2
poetry-core        1.1.0a6
```

**Comment by pythoninthegrass** [comment]
Updated bash to  5.1.8(1)-release. Re-ran the completion command and get this:
```bash
$ source ~/.bashrc
-bash: /usr/local/etc/profile.d/poetry.bash-completion: line 40: syntax error near unexpected token `clear'
-bash: /usr/local/etc/profile.d/poetry.bash-completion: line 40: `            (cache clear)'
```

**Comment by sgaist** [comment]
Did you check the value of `$BASH_VERSION` ?
IIRC, it's used by bash completion to load its helper functions and if it's to old it won't but without a warning.
An mismatched value can happen for the following reason: your login shell is still configured to point on the macOS provided bash and your "Shell opens with" Terminal setting is set to "Default login shell".

You have two options here:
1) Change "Shell opens with" to point to your brew bash
2) Change the default shell used when you login keeping the "Shell opens with" default value

For number 2, you need to add your brew bash path to the list of available shells in `/etc/shells` (requires root access) and then use `chsh -s <path/to/bash>` as your current user.

**Comment by sgaist** [comment]
From the looks of it, you did not install the `bash-completion` package.

Do that first and then put the poetry completion file under `/usr/local/etc/bash_completion.d/`.

**Comment by ablacklama** [comment]
I'm getting the same issue here on Ubuntu using the exact same commands and when i try to run it as a bash script normally. I have updated my bash and bash-completions to no avail. 

The problem seems to be that completions for subcommands are written like 
```bash
            (command subcommand)
            opts="${opts} ..."
            ;;
```
which isn't valid bash as far as i can tell. Separating the command and subcommand with a space doesn't seem to work.
But i'm no expert so take that with a grain of salt.

**Comment by abn** [comment]
I suspect this is an issue with `cleo` as we use it to handle the CLI ux.

**Comment by abn** [comment]
This one still fails accross all platforms.

```console
$ poetry --version
Poetry (version 1.2.0b2dev0)
$ echo $BASH_VERSION
5.1.16(1)-release
$ poetry completions bash | bash
bash: line 40: syntax error near unexpected token `clear'
bash: line 40: `            (cache clear)'
```

**Comment by koutoftimer** [comment]
Why this issue is in closed state? Isn't it a bug? Let it stay open until fixed, what is wrong with you?

**Comment by LordFckHelmchen** [comment]
Can we reopen this issue as it is still a problem and the PR that was supposed to close it, did actually do it (see there)?

**Comment by raffraffraff** [comment]
The poetry maintainers are playing whack-a-mole with duplicate tickets. Poetry users will probably keep creating more issues for this bug as long as the bug still exists in the latest version of poetry, and an open issue for the bug doesn't exist. Hand-waving at an upstream repo and closing all of the issues makes it extremely difficult to follow what's going on.

Suggestion: leave 1 issue open for this bug in poetry and link it to the root cause issue in cleo.

**Comment by neersighted** [comment]
I personally haven't looked at this issue for a minute (it merely lives in my 'common duplicates' doc) -- looks like it was closed by mistake as the Cleo PR should have specified Relates-To or similar. This shouldn't be closed in Poetry until after the Cleo release and subsequent bump of our constraint.

**Comment by raffraffraff** [comment]
If anyone wants working bash completion for poetry _now_, I created a gist: https://gist.github.com/raffraffraff/36b7a8cd1a4308150d7f843724a19ff9

I wanted an excuse to learn the poetry syntax, plus having decent shell completion makes learning it much easier.

YMMV, but it's working great for me. It's able to complete commands, positional args and options, and even suggests package names, source names, config options and caches where appropriate (and I don't think the official shell completion will do that).

**Comment by koutoftimer** [comment]
@raffraffraff without pull request it will be lost in the internet for sure. (I have fixed completions with vim+macros in <1 min).

**Comment by raffraffraff** [comment]
@koutoftimer I didn't send a pull request because poetry uses cleo for bash completions, but my script is 'hand made'. I'm just sharing here in case it's useful to others.

**Comment by lekjos** [comment]
FYI: I've got the same problem.

```
Poetry version: 1.2.2
Ubuntu version: 22.04.1 LTS (running in wsl)
Bash version: 5.1.16(1)-release
```

exact error is:

```
bash: /home/myusername/.bash_completion: line 40: syntax error near unexpected token `clear'
bash: /home/myusername/.bash_completion: line 40: `            (cache clear)'
```

**Comment by neersighted** [comment]
This is well-known and universal; please refrain from "me too" as it only adds noise and distracts the people trying to get Cleo 1.0 over the finish line so we can ship the (already merged) fix 😄

**Comment by Secrus** [comment]
Completion bugs are now fixed with the incoming release.

**Comment by ab** [comment]
When is version 1.3.0 expected to be published?

**Comment by neersighted** [comment]
When it's ready/the maintainers have had time to finalize the changelog (including any late merges).

**Comment by benkehoe** [comment]
If you go to the "watch" dropdown for this repository in the upper right corner of the page and click "custom", you can select to be notified of [releases](https://github.com/python-poetry/poetry/releases), which should send you an alert when 1.3.0 is published.

**Comment by pythoninthegrass** [comment]
Just wanted to close the loop here (know the _issue_ is closed haha) — running bash 5.2.15 and poetry 1.3.2 appears to have resolved the completion errors after re-exporting the completion script via
```bash
poetry completions bash > /usr/local/etc/profile.d/poetry.bash-completion
```
Appreciate everyone's help 🙏
